toplevel/genprintval.ml

```diff
@@ -500,6 +335,162 @@ module Make(O : OBJ)(EVP : EVALPATH with type valu = O.t) = struct
               Oval_stuff "<module>"
         end
 
+      and tree_of_list depth obj ty_arg =
+        if not (O.is_block obj) then Oval_list []
+        else match check_depth depth obj ty with
+          | Some x -> x
+          | None ->
+              let rec tree_of_conses tree_list depth obj ty_arg =
+                if !printer_steps < 0 || depth < 0 then
+                  Oval_ellipsis :: tree_list
+                else if O.is_block obj then
+                  let tree = nest tree_of_val (depth - 1)
+                                (O.field obj 0) ty_arg
+                  in
+                  let next_obj = O.field obj 1 in
+                  nest_gen (Oval_stuff "<cycle>" :: tree :: tree_list)
+                    (tree_of_conses (tree :: tree_list))
+                    depth next_obj ty_arg
+                else tree_list
+              in
+              Oval_list
+                  (List.rev (tree_of_conses [] depth obj ty_arg))
+
+      and tree_of_array depth obj ty_arg =
+        let length = O.size obj in
+        if length = 0 then Oval_array []
```

Comment:
The original check is `length <= 0`.

Reply:
I wasn't sure whether you are suggesting to use `length <= 0`, but I would find it weird given that values don't have a negative size. I propose to stick with `length = 0` which should be equivalent and clearer.

Reply:
Mostly a self-comment during my review.